### ISSUE TITLE
[FYI]: `-Wnomodule-import-in-extern-c` CCFLAGS is required for Xcode 15

### ISSUE THREAD

**Issue by r-plus** [theos/theos#791]
### What are the steps to reproduce this issue?

1. build tweak with Xcode 15

### What happens?

compile error from substrate header.

```
==> Compiling Tweak.xm (arm64e)…
In file included from Tweak.xm:38:
/opt/theos/vendor/include/substrate.h:45:1: error: import of C++ module 'MachO.nlist' appears within extern "C" language linkage specification [-Wmodule-import-in-extern-c]
#include <mach-o/nlist.h>
^
/opt/theos/vendor/include/substrate.h:43:1: note: extern "C" language linkage specification begins here
extern "C" {
^
1 error generated.
```

### What were you expecting to happen?

no error.

### Paste any relevant logs, error output, etc.

```shell
We should add `-Wnomodule-import-in-extern-c` to CCFLAGS to fixing this issue.

Ref: https://stackoverflow.com/questions/77203592/xcode-15-import-of-c-module-darwin-c-time-appears-within-extern-c-language
```

### What OS are you seeing the problem on?

macOS

### What OS version does it have installed?

14.3.1

### What toolchain and version are you using?

Xcode 15.1

### Which SDK version are you using?

automatic selection

### What OS is your client device running?

iOS

### What OS version does it have installed?

n/a

### Which device model is it?

n/a

### Which jailbreak is it using (if relevant)?

n/a

**Comment by L1ghtmann** [comment]
Hey, thanks for reporting this! May be worth just adjusting the header to avoid importing in extern. We have discussed this briefly in the Theos discord but don't believe we ever settled on the desired solution.

**Comment by leptos-null** [comment]
Discord discussion for reference: https://discord.com/channels/811490080278839327/811494949849661490/1188603122814296155

Adding to Lightmann's comment above:

This is a change in the iOS 17 SDK (unrelated to toolchain). Currently we suggest using the Theos SDKs instead of the Xcode SDK.
This module change is a breaking change, so not sure how to best support in Theos right now.

**Comment by zx500xl** [comment]
@leptos-null, how to use the Theos SDKs instead of the Xcode SDK? What do I have to write in the Makefile?

**Comment by leptos-null** [comment]
@zx500xl 
1. Download the theos SDKs you would like to use from https://github.com/theos/sdks/releases/tag/master-146e41f
2. Extract the archives and place the `.sdk`s in `${THEOS}/sdks/`. For example `${THEOS}/sdks/iPhoneOS16.5.sdk`
3. Set the SDK version component of your target in your `Makefile`. Generally this looks like `TARGET := iphone:clang:16.5:14.0` where `16.5` is the version of the SDK you selected

**Comment by Cregin** [comment]
> @zx500xl
> 
> 1. Download the theos SDKs you would like to use from https://github.com/theos/sdks/releases/tag/master-146e41f
> 2. Extract the archives and place the `.sdk`s in `${THEOS}/sdks/`. For example `${THEOS}/sdks/iPhoneOS16.5.sdk`
> 3. Set the SDK version component of your target in your `Makefile`. Generally this looks like `TARGET := iphone:clang:16.5:14.0` where `16.5` is the version of the SDK you selected

@leptos-null The problem still exists,Xcode 14.3.1

**Comment by MikePendo** [comment]
having same issue any progress on it?

**Comment by L1ghtmann** [comment]
@MikePendo Not yet, no. Though this is likely to be an issue moving forward, currently the issue is limited to newer sdks so it hasn't been a high priority. 

Regarding a fix, relying on suppressing the error is not great. It seems to make sense to adjust the header, but we do not handle the resources for substrate. @sbingner do you have an opinion as to next steps? We can edit our version of substrate's header but then it will be out of sync with your/Jay's upstream.

**Comment by sbingner** [comment]
They made `mach-o/nlist.h` c++? That's rude.  I also don't control the substrate header, that comes from Saurik but I'm not even sure how to fix this other than an ugly SDK version #define check... newer SDKs had the `extern "C"` code in the nlist file that would let it be imported outside of an `extern "C"` in the including file, but the older ones do not.

This is also odd because `nlist.h` is not c++ even on SDK 17.2 - but they put it "inside" a c++ module. ugh.

**Comment by leptos-null** [comment]
> They made `mach-o/nlist.h` c++? 

No, it sounds like it's just safe to import into C++.

Looking at the iOS 9 SDK (earliest in `theos/sdks`), `nlist.h` has been safe to import into C++ since at least then: https://github.com/theos/sdks/blob/ca52092676249546f08657d4fc0c8beb26a80510/iPhoneOS9.3.sdk/usr/include/mach-o/nlist.h#L306

Looking at this more though, I don't see anywhere in `CydiaSubstrate.h` that the declarations in `nlist.h` are used.

In conclusion: I believe we can simply remove the `extern "C"` around the `nlist.h` import.
If we want to be more aggressive, we can remove the `nlist.h` import entirely.
